torch/csrc/Module.cpp

```diff
@@ -841,6 +842,66 @@ PyObject* THPModule_setCheckSparseTensorInvariants(
   Py_RETURN_NONE;
 }
 
+// This function tries to check if the given object looks like an nn.Module
+// and if so if it has hooks installed on it.
+// It will return None if it doesn't know how to handle the given object and you
+// should check the fields manually (we could do it here but this is a POC and there
+// is little benefit at doing these full checks with the C API).
+// It will return True if all hooks field are dict and are empty.
+// It will return False if the hook fiels were found and are not empty.
+PyObject* THPModule_try_check_module_has_hooks(PyObject*,
+    PyObject* const* args,
+    Py_ssize_t nargs) {
+  TORCH_CHECK(nargs == 1, "_try_check_module_has_hooks takes only one argument");
+  PyObject* self = args[0];
+  // We don't check the type but we also don't assume this is an nn.Module!
+  PyTypeObject* self_t = self->ob_type;
+
+  // Module are always heap types
+  if (!(self_t->tp_flags & Py_TPFLAGS_HEAPTYPE)) {
+    Py_RETURN_NONE;
+  }
+
+  PyHeapTypeObject *self_ht = (PyHeapTypeObject *)self_t;
+
+  // No slots, we can't do fast check
+  if (!self_ht->ht_slots) {
+    Py_RETURN_NONE;
+  }
+
+  bool res = true;
+
+  // TODO: Read this from nn.Module at initialization time to avoid duplicate
+  static int slots_size = 4;
+  static const char* slots[] = {"_backward_hooks", "_backward_pre_hooks", "_forward_hooks", "_forward_pre_hooks"};
+
+  // Wrong number of slots
+  if (slots_size != PyTuple_GET_SIZE(self_ht->ht_slots)) {
+    Py_RETURN_NONE;
+  }
+
+  PyMemberDef *mp = self_t->tp_members;
+
+  for (int i; i < slots_size; ++i) {
```

Comment:
uninitialized i

Reply:
I will NOT land this! This is just a POC to show how you would write such a function and the perf you get.
I would expect that this gets added somewhere in Dynamo with the proper error checking for Dynamo code ;)